qualtran/bloqs/chemistry/thc.py

```diff
@@ -0,0 +1,411 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""SELECT and PREPARE for the molecular tensor hypercontraction (THC) hamiltonian"""
+from functools import cached_property
+from typing import Dict, Tuple
+
+import cirq
+import numpy as np
+from attrs import frozen
+from cirq_ft.algos.arithmetic_gates import LessThanEqualGate, LessThanGate
+from cirq_ft.algos.multi_control_multi_target_pauli import MultiControlPauli
+from cirq_ft.algos.select_swap_qrom import SelectSwapQROM
+from cirq_ft.linalg.lcu_util import preprocess_lcu_coefficients_for_reversible_sampling
+
+from qualtran import Bloq, BloqBuilder, Register, Signature, SoquetT
+from qualtran.bloqs.arithmetic import EqualsAConstant, GreaterThanConstant, ToContiguousIndex
+from qualtran.bloqs.basic_gates import Hadamard, Ry, Toffoli, XGate
+from qualtran.bloqs.on_each import OnEach
+from qualtran.bloqs.swap_network import CSwapApprox
+from qualtran.cirq_interop import CirqGateAsBloq
+
+
+def add_from_bloq_register_flat_qubits(
```

Comment:
are these two functions obviated by @tanujkhattar 's new/advanced interop features?

Reply:
Quite possibly, I need to merge and fix things.

Reply:
ok ping me when that's ready

Reply:
One of them was, the arithmetic gates still expect split registers

Reply:
@tanujkhattar can you comment here?

Reply:
xref #394 and #360

Reply:
Ok these are now fixed given the recent merges. I'll incorporate the new qualtran bloqs here before merging.